### DISCUSSION TITLE
How to add an extra encoder for transformer transducer?

### DISCUSSION THREAD

**Discussion by gyc666** [espnet/espnet#4169] [Q&A] [answered]
How to add an extra encoder 【an pretrained bert or BLSTM】 for transformer transducer?

How I modify the train.yaml and where should i modify on espnet.nets.pytorch_backend.e2e_asr_transducer:E2E ?

What should I pay attention to?

I would be most grateful if you could answer my questions！

**Comment by b-flo** [comment, marked as answer]
Hi,

> How to add an extra encoder 【an pretrained bert or BLSTM】 for transformer transducer?
> What should I pay attention to?

There are two "kind" of encoder architectures for Transducer in ESPnet1 (cf. [the documentation](https://espnet.github.io/espnet/tutorial.html#transducer)) : [RNNEncoder](https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/transducer/rnn_encoder.py) and [CustomEncoder](https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/transducer/custom_encoder.py), where the latter refer to the customizable architecture block-by-block, supporting Transformer/Conformer/Conv1d blocks.

Thus, I would suggest looking at the second class and how we [build](https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/transducer/blocks.py) the architecture.

> How I modify the train.yaml and where should i modify on espnet.nets.pytorch_backend.e2e_asr_transducer:E2E ?

Depending on what you want to do exactly, changes in `e2e_asr_transducer.py` may not be needed. For the changes in training config, well, it depends on your implementation!

Small note though: RNN-based and Transformer/Conformer-based encoders are separated. Mixing them together is not supported in this version. It can be done but bigger changes would be needed to do it "properly".

**Reply by gyc666** [reply]
Thanks very much ! I will try!
